agent/consul/stream/event_snapshot.go

```diff
@@ -9,50 +9,45 @@ package stream
 // collected automatically by Go's runtime. This simplifies snapshot and buffer
 // management dramatically.
 type eventSnapshot struct {
-	// Head is the first item in the buffer containing the snapshot. Once the
-	// snapshot is complete, subsequent BufferItems are appended to snapBuffer,
-	// so that subscribers receive all the events from the same buffer.
-	Head *bufferItem
+	// First item in the buffer. Used as the Head of a subscription, or to
+	// splice this snapshot onto another one.
+	First *bufferItem
 
-	// snapBuffer is the Head of the snapshot buffer the fn should write to.
-	snapBuffer *eventBuffer
+	// buffer is the Head of the snapshot buffer the fn should write to.
+	buffer *eventBuffer
 }
 
-// newEventSnapshot creates a snapshot buffer based on the subscription request.
-// The current buffer head for the topic requested is passed so that once the
-// snapshot is complete and has been delivered into the buffer, any events
-// published during snapshotting can be immediately appended and won't be
-// missed. Once the snapshot is delivered the topic buffer is spliced onto the
-// snapshot buffer so that subscribers will naturally follow from the snapshot
-// to wait for any subsequent updates.
-func newEventSnapshot(req *SubscribeRequest, topicBufferHead *bufferItem, fn SnapshotFunc) *eventSnapshot {
-	buf := newEventBuffer()
-	s := &eventSnapshot{
-		Head:       buf.Head(),
-		snapBuffer: buf,
+// newEventSnapshot creates an empty snapshot buffer.
+func newEventSnapshot() *eventSnapshot {
+	snapBuffer := newEventBuffer()
+	return &eventSnapshot{
+		First:  snapBuffer.Head(),
+		buffer: snapBuffer,
 	}
+}
 
-	go func() {
```

Comment:
I removed this goroutine as part of this change. Looking at `spliceFromTopicBuffer` I believe this method should not be long running. Either it immediately finds the empty item at the head, or it should pretty quickly traverse the buffer and find an appropriate item.

I guess the goroutine may be a small optimization in the case where the buffer is long, but we don't have any way to really control the lifecycle of the goroutine, so keeping it synchronous seems safer. This shouldn't be in the critical path of streaming events, because it is called from `Subscribe`, so adding up to a few microseconds to a subscribe call hopefully is not a problem.

Reply:
Yeah breaking it out of being quietly started in the background is good. We can always run `appendAndSlice in a go routine at the call site if we decide it's beneficial.

I think it might be for large snapshots especially coupled with the fact that this is no longer done under a lock so multiple callers might all be making snapshots at once and this is more likely the longer the snapshot takes now it is syncronous.

But if we solve the single flight issue then it doesn't matter whether the subsequent request waits for the lock due to a syncronous snapshot, or gets a pointer and then waits on the chan for the snapshot to be populated.